src/data/matrix/kronecker.lean

```diff
@@ -0,0 +1,319 @@
+/-
+Copyright (c) 2021 Filippo A. E. Nuccio. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Filippo A. E. Nuccio, Eric Wieser
+-/
+
+import data.matrix.basic
+import linear_algebra.tensor_product
+import ring_theory.tensor_product
+
+/-!
+# Kronecker product of matrices
+
+This defines the [Kronecker product](https://en.wikipedia.org/wiki/Kronecker_product).
+
+## Main definitions
+
+* `matrix.kronecker_map`: A generalization of the Kronecker product: given a map `f : α   → β → γ`
+  and matrices `A` and `B` with coefficients in `α` and `β`, respectively, it is defined as the
+  matrix with coefficients in `γ` such that
+  `kronecker_map f A B (i₁, i₂) (j₁, j₂) = f (A i₁ j₁) (B i₁ j₂)`.
+* `matrix.kronecker_map_linear`: when `f` is bilinear, so is `kronecker_map f`.
+
+## Specializations
+
+* `matrix.kronecker`: An alias of `kronecker_map (*)`. Prefer using the notation.
```

Comment:
I would call this `kronecker_prod` or `kronecker_mul` rather than simply `kronecker`.

Reply:
My only fear with `kronecker_mul` is that we end up with the lemma `mul_kronecker_mul_mul` which is confusing!

Reply:
Another option would be `kmul`

Reply:
Good option!

Reply:
@faenuccio but it is namespaced to `matrix.kronecker`. I think that's sufficiently unambiguous, right?